In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
# loading the dataset to a Pandas DataFrame
credit_card_data = pd.read_csv('/content/Dataset/creditcard.csv')

In [3]:
# first 5 rows of the dataset
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [4]:
credit_card_data.tail()

Time        V1        V2        V3        V4        V5        V6  \
31775  36510  1.497714 -0.874208  0.004261 -1.537368 -1.044513 -0.733905   
31776  36512 -0.407854  0.940336  1.533733 -0.018123 -0.219400 -0.934602   
31777  36513 -0.434226 -3.705556  0.194223  0.139216 -2.066818  1.220137   
31778  36513 -0.274278  1.004565  1.336430 -0.164777  0.404935 -0.418731   
31779  36513  1.295137  0.096775  0.348134  0.629896 -0.172700 -0.154350   

             V7        V8        V9  ...       V21       V22       V23  \
31775 -0.644635 -0.186510 -2.380495  ... -0.527761 -1.248016  0.145825   
31776  0.597172  0.024610 -0.151850  ... -0.247570 -0.725830 -0.006221   
31777 -0.278115  0.281452 -0.161302  ...  0.620431  0.040315 -0.690176   
31778  0.853252 -0.148772 -0.546969  ... -0.277437 -0.579135 -0.097793   
31779       NaN       NaN       NaN  ...       NaN       NaN       NaN   

            V24       V25       V26       V27       V28  Amount  Class  
31775 -0.061193  0.250252 -0.526898  0.003761  0.007950   24.90    0.0  
31776  0.322366 -0.249413  0.076587  0.255380  0.117340    1.79    0.0  
31777 -0.226844 -0.123369 -0.291412 -0.080514  0.155457  861.12    0.0  
31778 -0.037926 -0.157882  0.044814  0.078376 -0.137035    9.72    0.0  
31779       NaN       NaN       NaN       NaN       NaN     NaN    NaN  

[5 rows x 31 columns]

In [5]:
# dataset informations
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31780 entries, 0 to 31779
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    31780 non-null  int64  
 1   V1      31780 non-null  float64
 2   V2      31780 non-null  float64
 3   V3      31780 non-null  float64
 4   V4      31780 non-null  float64
 5   V5      31780 non-null  float64
 6   V6      31780 non-null  float64
 7   V7      31779 non-null  float64
 8   V8      31779 non-null  float64
 9   V9      31779 non-null  float64
 10  V10     31779 non-null  float64
 11  V11     31779 non-null  float64
 12  V12     31779 non-null  float64
 13  V13     31779 non-null  float64
 14  V14     31779 non-null  float64
 15  V15     31779 non-null  float64
 16  V16     31779 non-null  float64
 17  V17     31779 non-null  float64
 18  V18     31779 non-null  float64
 19  V19     31779 non-null  float64
 20  V20     31779 non-null  float64
 21  V21     31779 non-null  float64
 22

In [16]:
# checking the number of missing values in each column
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        1
V8        1
V9        1
V10       1
V11       1
V12       1
V13       1
V14       1
V15       1
V16       1
V17       1
V18       1
V19       1
V20       1
V21       1
V22       1
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [17]:

# dropping the missing values
credit_card_data = credit_card_data.dropna()
# checking the number of missing values in each column
credit_card_data.isnull().sum()


Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [18]:
# distribution of legit transactions & fraudulent transactions
credit_card_data['Class'].value_counts()

Class
0.0    31677
1.0      102
Name: count, dtype: int64

This Dataset is highly unblanced


0 --> Normal Transaction

1 --> fraudulent transaction

In [19]:
# separating the data for analysis
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]

In [20]:
print(legit.shape)
print(fraud.shape)

(31677, 31)
(102, 31)


In [21]:
# statistical measures of the data
legit.Amount.describe()

count    31677.000000
mean        81.082407
std        223.072655
min          0.000000
25%          6.870000
50%         20.000000
75%         73.610000
max       7879.420000
Name: Amount, dtype: float64

In [22]:
fraud.Amount.describe()

count     102.000000
mean       91.237451
std       248.270971
min         0.000000
25%         1.000000
50%         3.440000
75%        99.990000
max      1809.680000
Name: Amount, dtype: float64

In [23]:
# compare the values for both transactions
credit_card_data.groupby('Class').mean()

Time        V1        V2         V3        V4        V5  \
Class                                                                    
0.0    22335.407141 -0.188381  0.085855   0.759238  0.192821 -0.192581   
1.0    20334.872549 -7.761095  5.946853 -11.035230  5.926171 -5.529543   

             V6        V7        V8        V9  ...       V20      V21  \
Class                                          ...                      
0.0    0.095930 -0.095538  0.019413  0.324683  ...  0.041991 -0.03532   
1.0   -2.295415 -7.767215  3.904551 -2.997290  ...  0.679987  0.64190   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.118406 -0.039544  0.009212  0.133785  0.022140  0.010340  0.003013   
1.0   -0.352131 -0.317568 -0.241053  0.314412  0.177056  0.811372  0.118708   

          Amount  
Class             
0.0    81.082407  
1.0    91.237451  

[2 rows x 30 columns]

As data is very much imbalance so we use hybrid sampling under sample the class 0 to 10000 and over sample the class 1 to class 0

Build a sample dataset containing similar distribution of normal transactions and Fraudulent Transactions

Number of normal Transactions --> 31677


In [45]:
from imblearn.under_sampling import RandomUnderSampler
from imblearn.over_sampling import SMOTE
from collections import Counter

# Define features (X) and target (y)
X = credit_card_data.drop('Class', axis=1) # Replace with your actual target column name
y = credit_card_data["Class"]


# Scale the Features FIRST
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


minority_count = sum(y == 1)

# Step 1: Under-Sample the majority class (Reduce Class 0 to 10,000 samples)
under_sampler = RandomUnderSampler(sampling_strategy={0: minority_count * 10}, random_state=42)
X_under, y_under = under_sampler.fit_resample(X, y)

# Step 2: Over-Sample the minority class (Balance it with Class 0)
smote = SMOTE(sampling_strategy=1.0, random_state=42)  # Fully balance classes
X_balanced, y_balanced = smote.fit_resample(X_under, y_under)

# Check class distribution before & after balancing
print("Original Class Distribution:", Counter(y))  # Example: {0: 31000, 1: 102}
print("Balanced Class Distribution:", Counter(y_balanced))  # Example: {0: 10000, 1: 10000}

Original Class Distribution: Counter({0.0: 31677, 1.0: 102})
Balanced Class Distribution: Counter({0.0: 1020, 1.0: 1020})


In [46]:
new_dataset = pd.concat([X_balanced, y_balanced], axis=1)


In [47]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
0   3748 -0.052246 -0.036001  0.360702 -2.276312  1.996213  3.472714   
1   3682 -0.554882  1.290782  1.074192  0.104950  0.226145 -0.667515   
2   2339 -4.679468  3.892197  0.089051 -2.468071 -0.765521 -1.042147   
3  23607 -2.268397  0.009560  1.655149 -2.161494 -0.206615 -0.000516   
4  22164  1.185124 -0.383987  0.993934 -0.553815 -1.039601 -0.325390   

         V7        V8        V9  ...       V21       V22       V23       V24  \
0 -0.236639  0.570968 -0.093997  ... -0.252029 -0.667788 -0.003284  0.907996   
1  0.547436 -0.058588  1.263789  ... -0.491404 -1.076656 -0.040558 -0.294853   
2  0.948518 -0.510880  4.075118  ... -0.811750 -0.279780  0.040202  0.500288   
3 -0.468789  0.676074  0.134288  ... -0.643489 -1.237692  0.412480 -0.433970   
4 -0.789579  0.072944  3.237141  ... -0.138948  0.063306  0.006837  0.013049   

        V25       V26       V27       V28  Amount  Class  
0 -0.108657 -0.575422 -0.136576 -0.171139   39.99    0.0  
1 -0.091241  0.074699  0.320965  0.147209   11.99    0.0  
2  0.363647  0.480850 -0.315070 -1.110600    1.46    0.0  
3  0.000207  0.309524 -0.309244  0.129337   13.80    0.0  
4  0.369898 -0.693637  0.075486  0.025435   11.85    0.0  

[5 rows x 31 columns]

In [48]:
new_dataset.tail()

Time         V1         V2         V3        V4         V5        V6  \
2035  26559 -19.296805  11.889530 -22.034288  6.084344 -14.865777 -4.265639   
2036  28684 -29.050729  16.094069 -29.893142  6.470275 -21.128540 -4.894701   
2037  27232 -25.550704  14.440489 -27.052882  6.361507 -18.849217 -4.663134   
2038  29613   1.042400   1.990741  -3.577655  3.700275   0.928022 -1.409150   
2039    412  -2.381893   1.465272  -1.352805  3.835079  -0.342906 -1.392087   

             V7         V8        V9  ...       V21       V22       V23  \
2035 -15.327576  12.934293 -4.006236  ...  1.746844 -1.345441 -0.642297   
2036 -19.727029  19.075515 -3.622936  ...  1.808556 -2.163326 -1.354607   
2037 -18.093258  16.809523 -3.757523  ...  1.782221 -1.885056 -1.208196   
2038  -0.174097   0.239342 -1.552565  ...  0.232721  0.334061 -0.100952   
2039  -2.264657   1.252628 -2.532018  ...  0.530994  0.009774 -0.289818   

           V24       V25       V26       V27       V28     Amount  Class  
2035  0.110991  1.339142 -0.221725  1.610246  0.479078  99.990000    1.0  
2036  0.173308  2.092316 -0.210066  1.283803  0.374121  99.990000    1.0  
2037  0.158533  1.790141 -0.219682  1.402152  0.411356  99.990000    1.0  
2038 -0.237781  0.243315  0.161120  0.171051  0.084814  10.673158    1.0  
2039  0.261708  0.066932  0.147051  0.212188 -0.126220  50.393338    1.0  

[5 rows x 31 columns]

In [49]:
new_dataset['Class'].value_counts()

Class
0.0    1020
1.0    1020
Name: count, dtype: int64

In [50]:
new_dataset.groupby('Class').mean()

Time        V1        V2         V3       V4        V5  \
Class                                                                   
0.0    22431.135294 -0.191103  0.065256   0.792327  0.17156 -0.196521   
1.0    20575.582353 -8.169095  6.335403 -11.721757  6.19507 -5.842701   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.077275 -0.085604 -0.014082  0.401203  ...  0.065714 -0.054904   
1.0   -2.385328 -8.273249  4.211973 -3.200936  ...  0.733864  0.659577   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.107348 -0.037855  0.000488  0.146819 -0.008334  0.001301 -0.008829   
1.0   -0.357279 -0.279764 -0.295369  0.380429  0.202108  0.866431  0.137706   

          Amount  
Class             
0.0    80.955745  
1.0    69.930521  

[2 rows x 30 columns]

In [51]:
# Splitting the data into Features & Targets
X = new_dataset.drop(columns='Class', axis=1)
Y = new_dataset['Class']

In [52]:
X.shape

(2040, 30)

In [53]:
Y.shape

(2040,)

In [54]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

from sklearn.preprocessing import StandardScaler
# Initialize the scaler
scaler = StandardScaler()

# Fit on training data and transform both train & test
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [55]:
print(X.shape, X_train.shape, X_test.shape)

(2040, 30) (1632, 30) (408, 30)


In [56]:
model = LogisticRegression(max_iter=1000, solver='saga')

In [57]:
# training the Logistic Regression Model with Training Data
model.fit(X_train, Y_train)

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


LogisticRegression(max_iter=1000, solver='saga')

In [58]:
# accuracy on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [59]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.8651960784313726


In [60]:
# accuracy on test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [61]:
print('Accuracy score on Test Data : ', test_data_accuracy)

Accuracy score on Test Data :  0.8627450980392157


In [62]:


from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier


# Define the parameter grids for each model
param_grid_svm = {
    'C': [0.1, 1, 10],
    'kernel': ['linear', 'rbf'],
    'gamma': ['scale', 'auto']
}

param_grid_dt = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 5, 10],
    'min_samples_split': [2, 5, 10]
}

param_grid_rf = {
    'n_estimators': [100, 200],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5]
}


# Create model instances
models = {
    'SVM': SVC(),
    'Decision Tree': DecisionTreeClassifier(),
    'Random Forest': RandomForestClassifier()
}


# Iterate through models and perform grid search
for model_name, model in models.items():
    print(f"\nPerforming Grid Search for {model_name}:")

    if model_name == 'SVM':
        grid_search = GridSearchCV(model, param_grid_svm, cv=5, scoring='accuracy')
    elif model_name == 'Decision Tree':
        grid_search = GridSearchCV(model, param_grid_dt, cv=5, scoring='accuracy')
    elif model_name == 'Random Forest':
        grid_search = GridSearchCV(model, param_grid_rf, cv=5, scoring='accuracy')

    grid_search.fit(X_train_scaled, Y_train)

    print("Best Parameters:", grid_search.best_params_)
    print("Best Accuracy:", grid_search.best_score_)

    # Evaluate on the test set using the best model
    best_model = grid_search.best_estimator_
    test_accuracy = best_model.score(X_test_scaled, Y_test)
    print("Test Accuracy:", test_accuracy)



Performing Grid Search for SVM:
Best Parameters: {'C': 10, 'gamma': 'scale', 'kernel': 'rbf'}
Best Accuracy: 0.9963227706797246
Test Accuracy: 0.9975490196078431

Performing Grid Search for Decision Tree:
Best Parameters: {'criterion': 'entropy', 'max_depth': 10, 'min_samples_split': 2}
Best Accuracy: 0.9901896774919795
Test Accuracy: 0.9901960784313726

Performing Grid Search for Random Forest:
Best Parameters: {'max_depth': None, 'min_samples_split': 2, 'n_estimators': 100}
Best Accuracy: 0.9957055214723927
Test Accuracy: 0.9975490196078431


In [63]:


# Based on the provided code and the grid search results, the Random Forest model appears to be the best choice.
# Here's the code to train and evaluate the best model (Random Forest with optimized hyperparameters):
# Use the best parameters found from the grid search (adjust accordingly based on your results)
best_rf_params = {'max_depth': None, 'min_samples_split': 2, 'n_estimators': 200}

# Create and train the model with the best parameters
best_model = RandomForestClassifier(**best_rf_params)
best_model.fit(X_train_scaled, Y_train)

# Evaluate the model on the test set
test_accuracy = best_model.score(X_test_scaled, Y_test)
print("Test Accuracy (Best Random Forest):", test_accuracy)


Test Accuracy (Best Random Forest): 0.9975490196078431


In [99]:
!git add model/model.ipynb

fatal: pathspec 'model/model.ipynb' did not match any files
